src/main/scala/com/nec/vectorengine/VeProcessTraits.scala

```diff
@@ -0,0 +1,185 @@
+package com.nec.vectorengine
+
+import com.nec.colvector.{VeColVectorSource => VeSource}
+import scala.util.Try
+import java.nio.file.Path
+import org.apache.spark.api.plugin.PluginContext
+import org.bytedeco.javacpp.{LongPointer, Pointer}
+import org.bytedeco.veoffload.global.veo
+import org.bytedeco.veoffload.{veo_args, veo_proc_handle, veo_thr_ctxt}
+import com.typesafe.scalalogging.LazyLogging
+
+final case class LibraryReference private[vectorengine] (path: Path, value: Long)
+
+final case class LibrarySymbol private[vectorengine] (lib: LibraryReference, name: String, address: Long)
+
+final case class VeCallArgsStack private[vectorengine] (inputs: Seq[CallStackArgument], args: veo_args)
+
+sealed trait VeArgIntent
+
+object VeArgIntent {
+  final case object In extends VeArgIntent
+  final case object Out extends VeArgIntent
+  final case object InOut extends VeArgIntent
+}
+
+sealed trait CallStackArgument
+
+final case class I32Arg(value: Int) extends CallStackArgument
+
+final case class U64Arg(value: Long) extends CallStackArgument
+
+final case class BuffArg(intent: VeArgIntent, buffer: Pointer) extends CallStackArgument
+
+final case class VeAsyncReqId private[vectorengine] (value: Long)
+
+trait VeProcess {
+  def node: Int
+
+  def source: VeSource
+
+  def isOpen: Boolean
+
+  def apiVersion: Int
+
+  def version: String
+
+  def allocate(size: Long): Long
+
+  def free(location: Long): Unit
+
+  def put(buffer: Pointer): Long
+
+  def putAsync(buffer: Pointer): (Long, VeAsyncReqId)
+
+  def putAsync(buffer: Pointer, destination: Long): VeAsyncReqId
+
+  def get(buffer: Pointer, source: Long): Unit
+
+  def getAsync(buffer: Pointer, source: Long): VeAsyncReqId
+
+  /*
+    NOTE: `peek` here means "pick up a result from VE function if it has
+    finished", i.e. `peek and get`
+  */
+  def peekResult(id: VeAsyncReqId): (Int, LongPointer)
+
+  def awaitResult(id: VeAsyncReqId): LongPointer
+
+  def load(path: Path): LibraryReference
+
+  def unload(lib: LibraryReference): Unit
+
+  def getSymbol(lib: LibraryReference, symbol: String): LibrarySymbol
+
+  def newArgsStack(inputs: Seq[CallStackArgument]): VeCallArgsStack
+
+  def freeArgsStack(stack: VeCallArgsStack): Unit
+
+  def call(func: LibrarySymbol, stack: VeCallArgsStack): LongPointer
+
+  def callAsync(func: LibrarySymbol, stack: VeCallArgsStack): VeAsyncReqId
+
+  def close: Unit
+}
+
+object VeProcess extends LazyLogging {
+  final val DefaultVeNodeId = 0
+  final val MaxVeNodes = 8
+
+  private def createVeoTuple(venode: Int): Option[(Int, veo_proc_handle, veo_thr_ctxt)] = {
+    /*
+      If venode is -1, a VE process is created on the VE node specified by
+      environment variable VE_NODE_NUMBER. If venode is -1 and environment
+      variable VE_NODE_NUMBER is not set, a VE process is created on the VE
+      node #0.
+     */
+    val nnum = if (venode < -1) venode.abs else venode
```

Comment:
This looks to be a bit arbitrary, and being a comment on a private method is unlikely to end up in easily discoverable documentation. 

I think it probably makes sense to have that and any additional information about implicit behavior as scaladoc on createFromContext, so it sits right where one can do something about it.

Reply:
I will update this.

Reply:
This is now updated